### PR TITLE
Allow trailing commas for parameters, attribute arguments and method invocations

### LINKED ISSUE #8873: Allow trailing comma in parameter list
# Summary of the new feature/enhancement

Some languages like `Rust` allow a trailing comma in a parameter or argument list. In PowerShell, this would look like:

```powershell
param($a,)
```

Without allowing a trailing comma, some people resort to formatting like the following:

```powershell
param( [Parameter(Mandatory)][string] $Thing
     , [Parameter()][string] $AnotherThing
     , [Parameter()][string] $YetAnotherThing
)
```

The "leading" comma syntax gets used to minimize the lines changed when adding new parameters.

If a trailing comma was allowed, this style would be replaced with:

```powershell
param( [Parameter(Mandatory)][string] $Thing,
       [Parameter()][string] $AnotherThing,
       [Parameter()][string] $YetAnotherThing,
)
```

This could also be applied to method argument lists:

```powershell
$something.Insert($offset, $value,)
```

This usage might see less use than in parameter lists though as parameter lists usually span multiple lines whereas argument lists often do not.
```

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
fixes https://github.com/PowerShell/PowerShell/issues/8873
This will make trailing commas allowed in most scenarios:
```
param ($x,$y,)
[ValidateSet("X","Y",)]$x
"hello".Split('h','e',)
```
In all 3 cases, commas have also been made optional as long as the new argument isn't on the same line as the old one:
```
param (
    $x

    $y
)

[ValidateSet(
    "X"
    "Y"
)]$x

"hello".Split(
    'h'
    'e'
)
```

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Language/Parser/Parsing.Tests.ps1 [hunk 1]
```diff
@@ -644,3 +644,75 @@ Describe "Parsing array that has too many dimensions" -Tag CI {
...
+
+Describe "Trailing commas and no commas" -Tag CI {
+    It "Parse script succesfully: '<Script>'" -TestCases @(
+        @{ Script = @'
+param ($x
+$y)
+'@}
+        @{ Script = @'
+param ($x,$y,)
+'@}
```
Comment (iSazonov) [inline test/powershell/Language/Parser/Parsing.Tests.ps1:656]: I don't see tests like:
```c#
        @{ Script = @'
param ($x,
       $y,)
'@}
```
```c#
        @{ Script = @'
param ($x,
       $y,
)
'@}
```

File: test/powershell/Language/Parser/Parsing.Tests.ps1 [hunk 2]
```diff
@@ -644,3 +644,75 @@ Describe "Parsing array that has too many dimensions" -Tag CI {
...
+
+        $errs = Get-ParseResults -src $Script
+        $errs.Count | Should -Be $ErrorId.Count
+        for ($i = 0; $i -lt $errs.Count; $i++) {
+            $errs[$i].ErrorId | Should -BeExactly $ErrorId[$i]
+            $errs[$i].Extent.StartScriptPosition.Offset | Should -Be $StartOffset[$i]
+            $errs[$i].Extent.EndScriptPosition.Offset | Should -Be $EndOffset[$i]
+        }
+    }
+}
```
Comment (iSazonov) [inline test/powershell/Language/Parser/Parsing.Tests.ps1:718]:
```suggestion
}

```

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 3]
```diff
@@ -1216,6 +1225,7 @@ private AttributeBaseAst AttributeRule()
                             expr = new ConstantExpressionAst(name.Extent, true);
                             expressionOmitted = true;
                         }
+                        lastItemExtent = expr.Extent;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs:1228]:
```suggestion

                        lastItemExtent = expr.Extent;
```

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 4]
```diff
@@ -7917,6 +7932,11 @@ private List<ExpressionAst> InvokeParamParenListRule(Token lParen, out IScriptEx
                     {
                         if (comma != null)
                         {
+                            var endToken = PeekToken();
+                            if (endToken.Kind == TokenKind.RParen)
+                            {
+                                break;
+                            }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs:7939]:
```suggestion
                            }
                            
```

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 5]
```diff
@@ -925,34 +925,45 @@ private List<ParameterAst> ParameterListRule()
...
                 ParameterAst parameter = ParameterRule();
                 if (parameter == null)
                 {
-                    if (commaToken != null)
+                    if (nextToken is not null)
                     {
-                        // ErrorRecovery: ??
-
-                        ReportIncompleteInput(After(commaToken),
+                        var endToken = PeekToken();
```
Comment (vexx32) [inline src/System.Management.Automation/engine/parser/Parser.cs:936]: Do we need to define `endToken` as a separate variable here? Can we reuse `nextToken` and if not, why?

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 6]
```diff
@@ -1247,26 +1258,31 @@ private AttributeBaseAst AttributeRule()
...
-
-                        IScriptExtent errorExtent = After(commaToken);
-                        ReportIncompleteInput(
-                            errorExtent,
-                            nameof(ParserStrings.MissingExpressionAfterToken),
-                            ParserStrings.MissingExpressionAfterToken,
-                            commaToken.Kind.Text());
-                        positionalArguments.Add(new ErrorExpressionAst(errorExtent));
-                        lastItemExtent = errorExtent;
+                        var endToken = PeekToken();
```
Comment (vexx32) [inline src/System.Management.Automation/engine/parser/Parser.cs:1261]: Similar question here; does this need to be a separate variable -- can we reuse the existing `commaToken` (possibly renaming it), or is it significantly semantically different?

**Review by vexx32 — COMMENTED** [review]
Looks good overall to me, nice work!

**Comment by JamesWTruher** [comment]
i really don't care for this at all. it now means we have different parsing rules for parameters vs other instances of arrays which will be quite confusing. What problem is this trying to solve?

**Comment by vexx32** [comment]
There's another open issue to allow the same in array-literal `@($a, $b,)` syntax too I think, so this might just be one phase. Unsure if @MartinGC94 is planning on doing both changes though.

See: https://github.com/PowerShell/PowerShell/issues/16844

**Comment by PaulHigin** [comment]
A change of this scope should probably have an RFC first, to allow for discussion.  I am marking this for committee review.

**Comment by MartinGC94** [comment]
@vexx32 is right, I was inspired by https://github.com/PowerShell/PowerShell/issues/16844 . I looked into doing it for parentheses and array expressions as well but that looked difficult for me to do at first glance.  
Personally the feature I'm most interested in is removing the need for commas between parameters (this is also how they work in C#) and that's because it's always a mild annoyance to add/remove the commas when changing the parameters for a function. The rest were simply added because it looked like there was some interest so I figured "Why not?"

File: src/System.Management.Automation/engine/parser/Parser.cs (hunk 5 above)
Comment (MartinGC94) [reply src/System.Management.Automation/engine/parser/Parser.cs:936]: In both cases a new variable was used so I didn't change the behavior for the error reporting, which looks at the "nextToken". So I don't think it's strictly necessary but to keep changes minimal I decided to go with it.

**Comment by MartinGC94** [comment]
@SteveL-MSFT Was every individual part considered or just the whole change?  
Specifically I'm interested if making the commas optional in the param block could be reconsidered. It would be closer to how C# cmdlet parameters are defined, would be consistent with how array expressions work and if you don't close the param block with a parenthesis PSReadline should still prompt for input. Of course backwards compatibility would still be an issue, but I personally don't see why that should be weighed so high if the feature is inherently opt-in and any scripts written with 7.X in mind likely would need a lot of other changes to be made compatible with 5.1.

**Pull request closed without merging**